### PR TITLE
KAFKA-9651: Fix ArithmeticException (÷ by 0) in Partitioner impls

### PR DESCRIPTION
**Pull request by tombentley**
This PR fixes KAKFA-9651 by throwing UnknownTopicOrPartitionException instead of ArithmeticException.

### CODE REVIEW

**Comment by tombentley** [comment]
@cmccabe @mjsax @ewencp please could you review?

**Review by guozhangwang — COMMENTED** [review]
@tombentley In the producer code, we always block on metadata for that topic before calling `partition` so it is guaranteed that the cluster would never return empty list.

So the only problem is actually in DefaultStreamPartitioner: since we already have the num.partitions passed in as parameter, we can add a new overload function of `DefaultPartitioner#partition` that instead of taking a Cluster, we just take that num.partitions and then let DefaultStreamPartitioner to call that.

**Comment by tombentley** [comment]
Thanks @guozhangwang, I should have remembered that about the producer.

One thing I don't understand is how using `sourceTopicsInfo.maxPartitions` as the num partitions (as is done by `StreamsMetadataState.getKeyQueryMetadataForKey()`) will result in a non-zero number of partitions, when it appears to me that both code paths will use the same `Cluster` metadata (or maybe I'm wrong about that). 

I updated the commits if you want to take another look.

**Comment by guozhangwang** [comment]
`StreamPartitioner#partition`  is used in two places, in the RecordCollectorImpl the caller checks that numPartitions > 0; in the StreamsMetadataState as you asked, it is depending on `clusterMetadata.partitionsForTopic(topic);` indeed, but as long as KafkaStreams is running we are guaranteed that the source topics' metadata are available, and hence the maxPartitions is guaranteed to be non-zero.

**Comment by tombentley** [comment]
Thanks for the explanation.

**Comment by tombentley** [comment]
@guozhangwang any more comments on this PR?

**Comment by guozhangwang** [comment]
test this please

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java [hunk 1]
```diff
@@ -50,11 +48,29 @@ public void configure(Map<String, ?> configs) {}
      * @param cluster The current cluster metadata
      */
     public int partition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster) {
+        return partition(topic, key, keyBytes, value, valueBytes, cluster, 0);
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java]: minor: we can just calculate the `numPartitions` in this overload, and then in the other which has `numPartitions` passed in always blindly use that one.

**Review by guozhangwang — COMMENTED** [review]
Sorry for the late reply @tombentley . I left a minor comment, otherwise LGTM.

**Comment by tombentley** [comment]
Thanks @guozhangwang, now fixed.

**Pull request merged**